4. DSL Groovy

MZolot opened the pull request:
Начала работать над моделью и конфигурацией.

MZolot committed:
4. DSL Groovy
Начала работать над моделью и конфигурацией.

MZolot committed:
Конфигурация теперь нормально работает с дженериками.
Добавила класс для работы с гитом.

MZolot committed:
Добавила обработку исключений в работу с гитом.
Написала класс, отвечающий за сборку проектов.
Изменила некоторые классы модели, убрала лишние.
Начала работать над отчетом.

MZolot committed:
Сделала консольное приложение, которое выводит отчет в консоль.

molo4evan commented on Task_2_4_1/src/main/java/nsu/oop/handlers/GradleHandler.java:
```diff
@@ -0,0 +1,80 @@
+        test.withJvmTestClasses("*");
+        try {
+            test.run();
+        } catch (BuildException | TestExecutionException e) {
```
Есть вариант парсить сообщения эксепшнов, чтобы понять, какая именно ошибка произошла

MZolot replied:
это обязательно?

molo4evan commented on Task_2_4_1/src/main/java/nsu/oop/Commands.java:
```diff
@@ -0,0 +1,23 @@
+public class Commands implements Callable<Integer> {
+
+    @CommandLine.Option(names = "report", arity = "1", description = "configures the system and prints a report")
+    String repositoriesPath;
```
Это тоже стоит вынести в отдельный groovy-конфиг

MZolot replied:
что именно?..

molo4evan replied:
Я имел в виду, что путь к папке для репозиториев тоже имеет смысл указывать в groovy-конфиге.

molo4evan commented on Task_2_4_1/src/main/java/nsu/oop/reports/Report.java:
```diff
@@ -0,0 +1,85 @@
+        List<Student> students = groupConfig.getStudents();
+
+        RepositoryHandler repH = new RepositoryHandler(repositories);
+        repH.cloneRepository(students);
```
Мб стоит выполнять клонирование репозитория студента при его обработке? В текущем варианте ты будешь генерировать лишние ошибки, потому что все равно будешь пытаться собрать лабу в не склонированном репозитории.

molo4evan commented on Task_2_4_1/src/main/java/nsu/oop/reports/TaskReport.java:
```diff
@@ -0,0 +1,43 @@
+    public TaskReport (Task task, List<Student> students, String repositoriesPath) {
+        id = task.getId();
+        name = task.getName();
+        String folderName = "/Task_".concat(id);
```
А если студент назвал папку с задачей нестандартным образом?

MZolot replied:
Разве не было условия, что папки как раз (в отличие от веток) называются строго определенным образом?  Но могу и доделать, хорошо.

molo4evan commented on Task_2_4_1/src/main/java/nsu/oop/reports/TaskReport.java:
```diff
@@ -0,0 +1,43 @@
+        results = new ArrayList<>();
+        for (Student student : students) {
+            nsu.oop.handlers.GradleHandler gradleHandler = new nsu.oop.handlers.GradleHandler();
+            boolean buildRes = gradleHandler.buildTask(repositoriesPath.concat(student.getNickname()).concat(folderName));
```
Хотелось бы уметь проверять не только уже засчитанные таски (в мастере), но и те, которые еще не сданы (в соответствующих ветках). Стоит учитывать, что ветки могут тоже называться нестандартно.
При этом можно реализовать, например, поведение, при котором задача сначала ищется в мастере, а если не найдена, то в ветке из конфига.

molo4evan commented on Task_2_4_1/src/main/java/nsu/oop/model/Group.java:
```diff
@@ -0,0 +1,13 @@
+
+@EqualsAndHashCode(callSuper = true)
+@Data
+public class Group extends GroovyConfigurable {
```
Можешь добавить в репозиторий пример конфигов? Ну и для демонстрации работы будет полезно.

MZolot committed:
Добавила конфиги.
Добавила возможность указывать в конфиге на какой ветке и в какой папке лежит лаба. Запуск лаб из других папок работает, с других веток пока нет.

MZolot committed:
Добавила возможность проверять лабы с других веток

molo4evan commented on Task_2_4_1/src/main/java/nsu/oop/reports/TaskReport.java:
```diff
@@ -13,31 +18,84 @@ public class TaskReport {
+
+    private String findTaskFolder(String repositoriesPath, Student student) {
+        if (new File(repositoriesPath.concat(student.getNickname()).concat("/Task_").concat(id)).exists()) {
+            return repositoriesPath.concat(student.getNickname()).concat("/Task_").concat(id);
```
Вынеси в переменную ,чтобы дважды не конструировать

molo4evan commented on Task_2_4_1/src/main/java/nsu/oop/Main.java:
```diff
@@ -4,6 +4,7 @@
 
 public class Main {
     public static void main(String[] args) {
-        new CommandLine(new Commands()).execute(args);
```
Зачем хардкод? Если для тестовык запусков - лучше использовать Edit configuration в IDEA, а не переписывать код.

MZolot committed:
Добавила конфиг с настройками системы (путем к папке с репозиториями). Добавила возможность указывать в конфиге нужны ли тесты. Убрала хардкод в Main. Внесла мелкие правки.
